File: spring-data-requery/src/main/java/org/springframework/data/requery/converters/LocalDateTimeToLongConverter.java
package org.springframework.data.requery.converters;

import io.requery.Converter;

import javax.annotation.Nullable;
import java.time.LocalDateTime;
import java.time.ZoneOffset;

/**
 * org.springframework.data.requery.converters.LocalDateTimeToLongConverter
 *
 * @author debop
 */
public class LocalDateTimeToLongConverter implements Converter<LocalDateTime, Long> {


    @Override
    public Class<LocalDateTime> getMappedType() {
        return LocalDateTime.class;
    }

    @Override
    public Class<Long> getPersistedType() {
        return Long.class;
    }

    @Nullable
    @Override
    public Integer getPersistedSize() {
        return null;
    }

    @Override
    public Long convertToPersisted(LocalDateTime value) {
        return (value != null)
               ? value.toEpochSecond(ZoneOffset.UTC) * Java8Times.NANO_IN_SECONDS + value.getNano()
               : null;
    }

    @Override
    public LocalDateTime convertToMapped(Class<? extends LocalDateTime> type, @Nullable Long value) {
        if (value != null) {
            long epochSecond = value / Java8Times.NANO_IN_SECONDS;
            int nano = (int) (value % Java8Times.NANO_IN_SECONDS);

            return LocalDateTime.ofEpochSecond(epochSecond, nano, ZoneOffset.UTC);
        }

        return null;
    }
}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/converters/LocalDateTimeToLongConverter.kt
package org.springframework.data.requery.kotlin.converters

import io.requery.Converter
import java.time.LocalDateTime
import java.time.ZoneOffset

/**
 * Convert [LocalDateTime] to [Long]
 * @author debop
 * @since 18. 7. 2
 */
class LocalDateTimeToLongConverter: Converter<LocalDateTime, Long> {

    override fun getPersistedType(): Class<Long> = Long::class.java
    override fun getMappedType(): Class<LocalDateTime> = LocalDateTime::class.java
    override fun getPersistedSize(): Int? = null

    override fun convertToMapped(type: Class<out LocalDateTime>?, value: Long?): LocalDateTime? {
        return value?.let {
            val epochSecond = it / NANO_IN_SECONDS
            val nano = it % NANO_IN_SECONDS
            return LocalDateTime.ofEpochSecond(epochSecond, nano.toInt(), ZoneOffset.UTC)
        }
    }

    override fun convertToPersisted(value: LocalDateTime?): Long? {
        return value?.let {
            it.toEpochSecond(ZoneOffset.UTC) * NANO_IN_SECONDS + it.nano
        }
    }
}